src/Authoring/WinRT.SourceGenerator/AotOptimizer.cs

```diff
@@ -19,8 +19,28 @@ public class WinRTAotSourceGenerator : IIncrementalGenerator
     {
         public void Initialize(IncrementalGeneratorInitializationContext context)
         {
-            var properties = context.AnalyzerConfigOptionsProvider.Select(static (provider, _) => 
-                    (provider.IsCsWinRTAotOptimizerEnabled(), provider.IsCsWinRTComponent(), provider.IsCsWinRTCcwLookupTableGeneratorEnabled()));
+            var analyzerConfigProperties = context.AnalyzerConfigOptionsProvider.Select(static (provider, _) => (
+                    provider.IsCsWinRTAotOptimizerEnabled(),
+                    provider.IsCsWinRTComponent(),
+                    provider.IsCsWinRTCcwLookupTableGeneratorEnabled(),
+                    provider.IsCsWinRTAotOptimizerEnabledForBuiltInInterfaces(),
+                    provider.IsCsWinRTEmbeddedSupport())
+                );
+
+            var properties = context.CompilationProvider.Combine(analyzerConfigProperties).
+                    Select(static ((Compilation compilation, (bool isCsWinRTAotOptimizerEnabled, bool isCsWinRTComponent, bool, bool isCsWinRTAotOptimizerEnabledForBuiltInInterfaces, bool isCsWinRTEmbeddedSupport) properties) value, CancellationToken ct) =>
+                    // Enable the optimizer if the optimizer property is set and we would have a WinRT.Runtime reference.
+                    // There are a couple scenarios where we might not detect a WinRT.Runtime reference which we explicitly
+                    // check for such as component support where the generator introduces the reference and embedded support.
+                    // We will allow for an override for this by detecting whether CsWinRTAotWarningLevel is set to 2 which
+                    // indicates the component cares about optimization for built-in interfaces.
+                    (value.properties.isCsWinRTAotOptimizerEnabled && 
+                     (value.properties.isCsWinRTComponent ||
+                      value.properties.isCsWinRTAotOptimizerEnabledForBuiltInInterfaces ||
+                      value.properties.isCsWinRTEmbeddedSupport ||
+                      GeneratorHelper.HasWinRTRuntimeReference(value.compilation, ct)), 
+                     value.properties.isCsWinRTComponent,
+                     value.properties.Item3));

```

Comment:
Why are we returning the token..?

Reply:
which token?  Item3 is another bool which I am using to reconstruct the properties that is used later.

Reply:
Oh, sorry, misread. Too many tuple items ahah